runtime/compiler/microjit/SideTables.hpp

```diff
@@ -0,0 +1,294 @@
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#ifndef MJIT_SIDETABLE_HPP
+#define MJIT_SIDETABLE_HPP
+
+#include <cstdint>
+#include "microjit/x/amd64/AMD64Linkage.hpp"
+#include "microjit/utils.hpp"
+
+/*
+| Architecture 	| Endian | Return address | vTable index |
+| x86-64 		| Little | Stack 		  | R8 (receiver in RAX)
+
+| Integer Return value registers    | Integer Preserved registers   | Integer Argument registers
+| EAX (32-bit) RAX (64-bit)         | RBX R9                        | RAX RSI RDX RCX 
+
+| Float Return value registers  | Float Preserved registers | Float Argument registers
+| XMM0                          | XMM8-XMM15        		| XMM0-XMM7 
+*/
+namespace MJIT {
+
+struct RegisterStack {
+    unsigned char useRAX : 2;
+    unsigned char useRSI : 2;
+    unsigned char useRDX : 2;
+    unsigned char useRCX : 2;
+    unsigned char useXMM0 : 2;
+    unsigned char useXMM1 : 2;
+    unsigned char useXMM2 : 2;
+    unsigned char useXMM3 : 2;
+    unsigned char useXMM4 : 2;
+    unsigned char useXMM5 : 2;
+    unsigned char useXMM6 : 2;
+    unsigned char useXMM7 : 2;
+    unsigned char stackSlotsUsed : 8;
+};
+
+inline U_16
+calculateOffset(RegisterStack* stack)
+{
+    return  8*(stack->useRAX  +
+               stack->useRSI  +
+               stack->useRDX  +
+               stack->useRCX  +
+               stack->useXMM0 +
+               stack->useXMM1 +
+               stack->useXMM2 +
+               stack->useXMM3 +
+               stack->useXMM4 +
+               stack->useXMM5 +
+               stack->useXMM6 +
+               stack->useXMM7 +
+               stack->stackSlotsUsed);
+}
+
+inline void
+initParamStack(RegisterStack* stack)
+{
+    stack->useRAX         = 0;
+    stack->useRSI         = 0;
+    stack->useRDX         = 0;
+    stack->useRCX         = 0;
+    stack->useXMM0        = 0;
+    stack->useXMM1        = 0;
+    stack->useXMM2        = 0;
+    stack->useXMM3        = 0;
+    stack->useXMM4        = 0;
+    stack->useXMM5        = 0;
+    stack->useXMM6        = 0;
+    stack->useXMM7        = 0;
+    stack->stackSlotsUsed = 0;
+}
+
+inline int
+addParamIntToStack(RegisterStack* stack, U_16 size)
+{
+    if(!stack->useRAX){
+        stack->useRAX = size/8;
```

Comment:
What is the range of values for `size`? If it might be as large as 64, then `useRAX` must be wider than 2 bits.

Reply:
Thank you for the review Keith. This function is called from a function which is checking the parameters that are passed via type signature. It's converting bytes to slots so the range should be {8,16}. Though I am interested in knowing if there are any special cases which might not conform to that on the JIT side.

Reply:
Thank you for the review Keith. This function is called from a function which is checking the parameters that are passed via type signature. It's converting bytes to slots so the range should be {8,16}. Though I am interested in knowing if there are any special cases which might not conform to that on the JIT side.

Reply:
I'm not sure the space savings here is worth the added complexity: Why not just use `uint8_t` for fields like `useRAX`? The name could be improved the 'use' prefix implies 'boolean' to me.